introduced force_2d for a subset of geometries #180

whisk opened the pull request:
Added `force_2d` function for Point, MultiPoint, LineString, Polygon and all descending classes (if any). No support for MultiLineString and MultiPolygon yet.

pep8speaks commented:
Hello @whisk! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pygeoif/factories.py`](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/pygeoif/factories.py):

> [Line 88:90](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/pygeoif/factories.py#L88): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 89 characters)

* In the file [`tests/test_factories.py`](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py):

> [Line 30:22](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L30): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to False should be 'if cond is False:' or 'if not cond:'
> [Line 37:22](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L37): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to False should be 'if cond is False:' or 'if not cond:'
> [Line 46:1](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L46): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 48:5](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L48): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'
> [Line 54:5](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L54): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'
> [Line 59:1](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L59): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 80:22](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L80): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to False should be 'if cond is False:' or 'if not cond:'
> [Line 96:22](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L96): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to False should be 'if cond is False:' or 'if not cond:'
> [Line 100:59](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L100): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 100:90](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L100): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 89 characters)
> [Line 101:57](https://github.com/cleder/pygeoif/blob/b560ee9253104845489f558a35127065cc382331/tests/test_factories.py#L101): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

cleder commented:
Hi @whisk, thanks for your contribution.

I think that the `force_2d` and `force_3d` could be cleaner to implement by operating on the `__geo_interface__`  like [pygeoif.functions.compare_geo_interface](https://github.com/cleder/pygeoif/blob/8d349699af24931c7e465328d6c11bd0b98187af/pygeoif/functions.py#L166)

`def force_2d(geometry: Union[GeoType, GeoCollectionType]) -> Union[Geometry, GeometryCollection]:`

For `force_2d` you can [zip](https://docs.python.org/3.3/library/functions.html#zip) the coordinates with `(0 ,0)`, 

`def force_3d(geometry: Union[GeoType, GeoCollectionType], z: float = 0) -> Union[Geometry, GeometryCollection]:`

For `force_3d` you could use [zip_longest](https://docs.python.org/3/library/itertools.html#itertools.zip_longest) using `z` as the `fillvalue`

To convert the resulting `__geo_interface__` dictionary back into a geometry, you may use [shape](https://github.com/cleder/pygeoif/blob/8d349699af24931c7e465328d6c11bd0b98187af/pygeoif/factories.py#L100)

I hope this makes sense.

cleder commented:
https://realpython.com/python-zip-function/ and https://ioflood.com/blog/python-zip-function/ seem to be useful to learn more about `zip`

cleder commented:
Approach I'd take:

```
def move_coordinate(coordinate: Sequence[float], move_by: Sequence[float]) -> Tuple[float, ...]:
    if len(coordinate) > len(move_by):
        return tuple(c + m for c, m in zip(coordinate, move_by))
    else:
        return tuple(c + m for c, m in zip_longest(coordinate, move_by, fillvalue=0.0))
```
```
>>> move_coordinate((0,0), (0,0,0))
(0, 0, 0.0)
>>> move_coordinate((0,0), (0,0,1))
(0, 0, 1.0)
>>> move_coordinate((0,0,0), (0,0,1))
(0, 0, 1)
>>> move_coordinate((0,0,0), (0,0))
(0, 0)
```
This can be used by force_2d with `(0, 0)` as `move_by` and force_3d with `(0, 0, z)` as move_by

whisk commented:
Hi, I refactored `force_2d` by using `__geo_interface__`, like functions `compare_geo_interface` and `shape` do. This is much simpler and I think that's what you were looking for.

I also added `move_coordinate` and `move_coordinates` functions.

whisk commented on pygeoif/functions.py:
```diff
@@ -188,11 +190,60 @@ def compare_geo_interface(
+    >>> move_coordinates(((0, 0), (-1, 1)), (-1, 1, 0))
+    ((-1, 1, 0), (-2, 2, 0))
+    """
+    if isinstance(coordinates, (tuple, list)) and isinstance(
```
I preferred to check types explicitly rather what catch TypeError exception.

whisk replied:
Thanks for pointing this out!
As far as I see, you refactored this lines to use `CoordinatesType`, not `Sequence[Any]`, so this function just can't accept generators anyway. 
Not sure if generators still need to be supported in `move_coordinates`?

cleder replied:
You still CAN pass in a generator, it is just that mypy will complain that you should not do that

cleder replied:
Thanks again for your help, your work will be in the next release (1.2) :+1:

whisk replied:
> You still CAN pass in a generator, it is just that mypy will complain that you should not do that

Yes, yes, you are right. I'm probably just too used to strict typing :)
Please consider PR #188 regarding it.

cleder reviewed (approved):
Nice work :-) Thanks a lot!
